### PR TITLE
gh/workflows: Add datapath conformance suite

### PR DESCRIPTION
**Pull request by brb**
This commit introduces the GH action "ci-datapath". The action is used to run the cilium-cli's datapath connectivity suite from \[1\]. Currently, it is running on a single VM with a single configuration as an early proof whether the "ci-datapath" pipeline is going to be sound.

The new action introduces two novelties when compared to the existing GH actions in Cilium.

Firstly, we are using the "self-hosted" runner. This is handled by \[2\]. In short, for each instance of the action, the \[2\] project creates a VM which will run the action steps.

Secondly, we are using the "brb/lvh-action" \[3\] (TODO: to move to either isovalent/ or cilium/) which creates a LVH-based lightweight VM \[4\] on which the test cmds are being run.

The next steps are the following:
- Keep extending the datapath connectivity suite \[5\]
- Create a workflow matrix for different configurations and kernels, and introduce some parallelism (the "self-hosted" runner VM is capable of running multiple instances of LVH VMs at the same time).

\[1\]: https://github.com/cilium/cilium-cli/blob/master/connectivity/suite.go
\[2\]: https://github.com/isovalent/poor-mans-ephemeral-github-action-runner
\[3\]: https://github.com/brb/lvh-action
\[4\]: https://github.com/cilium/little-vm-helper
\[5\]: https://github.com/cilium/cilium/issues/20606

---

Previously the CI has passed - https://github.com/cilium/cilium/runs/8115113009.

Related #20606

### CODE REVIEW

File: .github/workflows/conformance-datapath.yaml
Comment (aanm) [inline .github/workflows/conformance-datapath.yaml]: We should probably not use a dependency that does not seem to be maintained nor is too well know. Copying a binary from a docker image is simple as doing:
```
container_id=$(docker create "quay.io/cilium/cilium-cli-ci:latest")
docker cp "$container_id:/usr/local/bin/cilium" "/usr/local/bin/cilium"
docker rm "$container_id"
```

File: .github/workflows/conformance-datapath.yaml
Comment (aanm) [inline .github/workflows/conformance-datapath.yaml]: Why `:latest`? shouldn't we pin a particular version? As it is done in all other GH workflows?

File: .github/workflows/conformance-datapath.yaml
Comment (aanm) [inline .github/workflows/conformance-datapath.yaml]: `brb/lvh-action@22b1c3623c9b64a294c20e1ac831b28850d33940`

Comment (aanm) [inline .github/workflows/conformance-datapath.yaml]: Similar to the question left regarding `quay.io/cilium/cilium-cli-ci:latest`. the same applies to the images in the lvh-action repository where it is using `quay.io/lvh-images/lvh:latest`.

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: I want to avoid creating docker container for that. The `undocker` is maintained and does the job (it's created by my friend). Or are you referring to `shrink/actions-docker-extract`?

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: See the code comment above ;-)

```
          # Also, we rely on the ":latest" since the datapath conformance suite
          # is heavily WIP.
```

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: The `lvh` is also very WIP. So for now I'd stick to the latest version.

Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: Why commit vs tag?

File: .github/workflows/conformance-datapath.yaml
Comment (aanm) [reply .github/workflows/conformance-datapath.yaml]: I'm referring to shrink/actions-docker-extract

File: .github/workflows/conformance-datapath.yaml
Comment (aanm) [reply .github/workflows/conformance-datapath.yaml]: tags are mutable

File: .github/workflows/conformance-datapath.yaml
Comment (joestringer) [reply .github/workflows/conformance-datapath.yaml]: I think I'm fine with `latest` in the intermediate period but I would keep this job marked "optional", ie *not* blocking PRs until this is stable. If you want to mark this target "required" to merge PRs, then let's get it into a stable state and set this to a fixed tag first.

File: .github/workflows/conformance-datapath.yaml
Comment (joestringer) [inline .github/workflows/conformance-datapath.yaml]: This should probably point to a sha at the very least, though if you're planning to replace it then should we fix that before merging this PR or after? It's always easy to defer tasks like that indefinitely if we don't push back at merge time.

File: .github/workflows/conformance-datapath.yaml
Comment (joestringer) [reply .github/workflows/conformance-datapath.yaml]: `cilium/little-vm-helper-action@....`?

File: .github/workflows/conformance-datapath.yaml
Comment (joestringer) [inline .github/workflows/conformance-datapath.yaml:192]: What's the high level of our plan to test Cilium integration with kube-proxy?

File: .github/workflows/conformance-datapath.yaml
Comment (joestringer) [inline .github/workflows/conformance-datapath.yaml]: @ sha

**Review by joestringer — CHANGES_REQUESTED** [review]
Seems fine to me from a @cilium/github-sec perspective. Just a few notes around using shas for action references and I have a couple of questions also below.

File: .github/workflows/conformance-datapath.yaml
Comment (christarazi) [inline .github/workflows/conformance-datapath.yaml]:
```suggestion
            mkdir -p cilium-sysdumps
```

```
MKDIR(1)                                           User Commands                                           MKDIR(1)

NAME
       mkdir - make directories

SYNOPSIS
       mkdir [OPTION]... DIRECTORY...

DESCRIPTION
       Create the DIRECTORY(ies), if they do not already exist.

       Mandatory arguments to long options are mandatory for short options too.
...

       -p, --parents
              no  error  if existing, make parent directories as needed, with their file modes unaffected by any -m
              option.
```

File: .github/workflows/conformance-datapath.yaml
Comment (christarazi) [inline .github/workflows/conformance-datapath.yaml:214]: Not sure if this was already discussed before, but we are running this suite on `master` 4 times a day and we are keeping 5 days worth of sysdumps. If a failure occurs  in this testing suite, who is going to triage the failures? Is it `@cilium/sig-datapath`?

File: .github/workflows/conformance-datapath.yaml
Comment (christarazi) [inline .github/workflows/conformance-datapath.yaml]:
```suggestion
          description: Datapath test failed
```

just to be consistent with the previous job's description.

File: .github/workflows/conformance-datapath.yaml
Comment (christarazi) [inline .github/workflows/conformance-datapath.yaml]: Same here re: consistency.

**Review by pchaigno — COMMENTED** [review]
Should we add this new file to CODEOWNERS?

File: .github/workflows/conformance-datapath.yaml
Comment (nbusseneau) [reply .github/workflows/conformance-datapath.yaml]: +1, best to avoid third party actions when possible.

File: .github/workflows/conformance-datapath.yaml
Comment (nbusseneau) [reply .github/workflows/conformance-datapath.yaml]: > Why commit vs tag?

Tags are not safe from being changed via force push (even if in this case we trust you, the author, not to be a mean person :D), so as part of our supply chain security policy we try to ensure all GitHub Actions steps and all container images be pinned to specific SHA.

File: .github/workflows/conformance-datapath.yaml
Comment (nbusseneau) [reply .github/workflows/conformance-datapath.yaml:214]: TopHat is supposed to triage scheduled runs, but honestly we should just have automation tools for scheduled runs. (I plan on working on that after I'm done with the current CI rework I'm prototyping on the side.)

File: .github/workflows/conformance-datapath.yaml
Comment (nbusseneau) [reply .github/workflows/conformance-datapath.yaml]: (nit: also use `tests` for consistency :D)

File: .github/workflows/conformance-datapath.yaml
Comment (nbusseneau) [inline .github/workflows/conformance-datapath.yaml]: Following merges of #21239 and #21238, I would say we drop this step and instead edit `.github/workflows/loki.yaml` as part of this PR.

**Review by nbusseneau — COMMENTED** [review]
Good stuff, I added some comments to the already existing discussions + a new one but overall LGTM.

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: Yep, the idea is to keep for now the DP action as optional until we it becomes stable and versioning is properly applied.

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: :+1: Moved to https://github.com/cilium/little-vm-helper-action

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml]: Replaced with @aanm's hack

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml:214]: > Is it @cilium/sig-datapath?

For now it's me, and once we are happy with the job, then the sig-datapath.

File: .github/workflows/conformance-datapath.yaml
Comment (tklauser) [inline .github/workflows/conformance-datapath.yaml:157]: This could now be switched to use cilium-cli 0.12.3 which contains the datapath connectivity tests.

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml:192]: Yep, the PR is just to demonstrate abilities of the CI pipeline. The test cases and configurations are listed in https://github.com/cilium/cilium/issues/20606.

**Comment by brb** [comment]
@pchaigno @christarazi @nbusseneau @joestringer @aanm Thanks for the review! I addressed your comments. Also, I moved the custom action to https://github.com/cilium/little-vm-helper-action/blob/main/action.yaml (@nbusseneau could you check it too?). The job with the new changes passed here https://github.com/cilium/cilium/actions/runs/3045574097/jobs/4907319486. PTAL.

File: .github/workflows/conformance-datapath.yaml
Comment (brb) [reply .github/workflows/conformance-datapath.yaml:157]: Thanks. However, I am adding more tests to the cilium-cli, so I don't want to depend on a specific version until we have a decent coverage.

**Review by nbusseneau — APPROVED** [review]
LGTM. Reviewed https://github.com/cilium/little-vm-helper-action/blob/main/action.yaml also, simple and sound, would only recommend pinning to a SHA for `actions/cache` but otherwise all good.

File: .github/workflows/conformance-datapath.yaml
Comment (tklauser) [reply .github/workflows/conformance-datapath.yaml:157]: Fair enough. But please be aware that pinning to `:latest` bears the risk of introducing unrelated flakes on this workflow due to flaky tests being merged to cilium-cli `master`.

Comment (brb) [reply .github/workflows/conformance-datapath.yaml:157]: :+1: Yep, this job is going to be "optional" in the beginning

**Comment by brb** [comment]
> would only recommend pinning to a SHA for actions/cache but otherwise all good.

@nbusseneau :+1: Changed - https://github.com/cilium/little-vm-helper-action/commit/b1ff6cddb67545973471cadc7a1620b71728c77e.

**Review by joestringer — APPROVED** [review]
I didn't look too closely but it looked like @nbusseneau already did. My feedback was addressed :+1:

**Review by aanm — APPROVED** [review]

**Review by christarazi — APPROVED** [review]

**Review by NikAleksandrov — APPROVED** [review]

**Pull request merged**
